### PR TITLE
feat: upgrade metrics server to v0.3.4

### LINKED ISSUE #1172: Upgrade metrics-server to v0.3.x
The new version of [metrics-server](https://github.com/kubernetes-incubator/metrics-server/releases) is supported on Kubernetes 1.8.x+ but has significant changes compared to the previous minor release.

See #1109.

### PR DESCRIPTION
**Pull request by andyzhangx**
**Reason for Change**:

This PR upgrades metrics server from `v0.2.1` to `v0.3.4`

BTW, we cannot disable `--read-only-port` now since AKS also depends on this port and metrics server on AKS is still v0.2.x

metric server(v0.2.1) is still using KubeletReadOnlyPort(10255), we cannot switch to KubeletPort(10250) now, after upgrade to metric server v0.3.4, it uses 10250 port instead, then we could move on to merge PR [fix: disable ReadOnlyPort ](https://github.com/Azure/aks-engine/pull/975)

This PR only applies for k8s v1.16.0:
 - upgrade metric server to v0.3.4

Tested on k8s v1.16.0-beta.1, test result:
```
$ kubectl top no
NAME                        CPU(cores)   CPU%   MEMORY(bytes)   MEMORY%
6133k8s010                  56m          2%     445Mi           8%
k8s-linuxpool1-61339693-0   54m          2%     692Mi           11%
k8s-master-61339693-0       92m          4%     1065Mi          17%
```

**Issue Fixed**:

Fixes #1172

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

cc @feiskyer 
/azp run pr-e2e

### CODE REVIEW

**Comment by andyzhangx** [comment]
/hold 
need to adjust the metrics server parameters to make it totally work

**Comment by zhiweiv** [comment]
Metrics server 0.3.x doesn’t work with Windows nodes yet, 
kubernetes/kubernetes#75934
kubernetes/kubernetes#76740

**Comment by andyzhangx** [comment]
> Metrics server 0.3.x doesn’t work with Windows nodes yet,
> [kubernetes/kubernetes#75934](https://github.com/kubernetes/kubernetes/issues/75934)
> [kubernetes/kubernetes#76740](https://github.com/kubernetes/kubernetes/issues/76740)

thanks for reminder, I think we can upgrade to Metrics server 0.3.0 on Linux node first.

**Comment by jackfrancis** [comment]
I would suggest we introduce this as a 1.15 feature and not backport to pre-existing version (and thus, pre-existing clusters built on pre-1.15).

If we do want to introduce this to older versions, we'll need thorough upgrade tests.

File: parts/k8s/containeraddons/kubernetesmasteraddons-metrics-server-deployment.yaml [hunk 1]
```diff
@@ -126,7 +126,6 @@ spec:
         imagePullPolicy: IfNotPresent
         command:
         - /metrics-server
-        - --source=kubernetes.summary_api:''
```
Comment (andyzhangx) [inline parts/k8s/containeraddons/kubernetesmasteraddons-metrics-server-deployment.yaml]: @jackfrancis how do we handle command parameter change in kubernetesmasteraddons_xxx.yaml file in aks-engine?

I am planning to upgrade `metrics-server` from v0.2.0 to v0.3.0 in k8s v1.15.0, while that require command parameters change, do you know what's the correct way to do this? One way could be move original `kubernetesmasteraddons-metrics-server-deployment.yaml` file to new folder like `1.14`, `1.13`, `1.12`, ... , looks like I need to create a lot of such version folder, is that the correct way?

Comment (jackfrancis) [reply parts/k8s/containeraddons/kubernetesmasteraddons-metrics-server-deployment.yaml]: Correct, the current way to define k8s version-specific changes is to create a new directory with the minor.major version of the first version that these changes work for. If there is only a 1.15 directory, for example (and not a 1.16 directory), then the logic assumes the spec is valid for 1.15 and above. So every time there is a forward looking version-breaking change, we create a new dir to store those with the same name as the version that introduces the change.

Hope that makes sense!

**Comment by andyzhangx** [comment]
Finally I have time to complete this PR, this PR only applies for k8s v1.15.0:
 - upgrade metric server to v0.3.1
 - disable `ReadOnlyPort` in kubelet

**Comment by andyzhangx** [comment]
/hold
Let's hold for now since for metrics server 0.3.x, it could not get CPU metrics due to issue: https://github.com/kubernetes/kubernetes/issues/75934

```
$ kubectl top nodes
NAME                    CPU(cores)   CPU%   MEMORY(bytes)   MEMORY%
8538k8s000              0m           0%     399Mi           7%
8538k8s001              0m           0%     421Mi           8%
k8s-master-85384734-0   112m         5%     1243Mi          20%
```

**Comment by andyzhangx** [comment]
not sure whether this test failure is related:
```
 /__w/1/s/gopath/src/github.com/Azure/aks-engine/test/e2e/kubernetes/kubernetes_test.go:157
------------------------------‌
S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌

Summarizing 1 Failure:‌

[Fail] ‌Azure Container Cluster using the Kubernetes Orchestrator ‌regardless of agent pool type ‌[It] should validate host OS DNS ‌
&#x2F;__w&#x2F;1&#x2F;s&#x2F;gopath&#x2F;src&#x2F;github.com&#x2F;Azure&#x2F;aks-engine&#x2F;test&#x2F;e2e&#x2F;kubernetes&#x2F;kubernetes_test.go:157‌

Ran 1 of 47 Specs in 912.101 seconds‌
FAIL!‌ -- ‌0 Passed‌ | ‌1 Failed‌ | ‌0 Pending‌ | ‌46 Skipped‌
--- FAIL: TestKubernetes (912.10s)
FAIL
```

**Comment by jackfrancis** [comment]
@andyzhangx I rebased this against master today and re-ran E2E: getting the runAsRoot error described here in the test runs:

https://github.com/Azure/aks-engine/issues/1851

Any ideas there?

**Comment by PierluigiLenociAkelius** [comment]
https://github.com/kubernetes-incubator/metrics-server/issues/272

File: pkg/api/k8s_versions.go [hunk 2]
```diff
@@ -19,7 +19,7 @@ var k8sComponentVersions = map[string]map[string]string{
 		"exechealthz":                      "exechealthz-amd64:1.2",
 		"addon-resizer":                    "addon-resizer:1.8.5",
 		"heapster":                         "heapster-amd64:v1.5.4",
-		"metrics-server":                   "metrics-server-amd64:v0.2.1",
+		"metrics-server":                   "metrics-server-amd64:v0.3.1",
```
Comment (PatrickLang) [inline pkg/api/k8s_versions.go]: will change to 0.3.4

Comment (PatrickLang) [reply pkg/api/k8s_versions.go]: done

**Comment by PatrickLang** [comment]
Windows 1.16 tests failed the first attempt for unrelated reasons. Restarted it.
"Allocation failed. We do not have sufficient capacity for the requested VM size in this region. Read more about improving likelihood of allocation success at http://aka.ms/allocation-guidance

**Comment by jackfrancis** [comment]
Rekicked the tests, though the HPA failure in Linux test is almost certainly related to metrics-server non-functionality

**Comment by jackfrancis** [comment]
Peeking at the E2E tests in progress: it looks like the repro is HPA doesn't scale pods back down.

**Comment by PatrickLang** [comment]
Autoscaler tests are failing on Linux:

```

    ‌Expected error:‌
        <*errors.fundamental | 0xc0003fc060>: {
            msg: "Timeout exceeded (20m0s) while waiting for minimum -1 and maximum 1 Pod replicas from Deployment php-apache-long-running",
            stack: [0x8e64a9, 0x45e861],
        }
        Timeout exceeded (20m0s) while waiting for minimum -1 and maximum 1 Pod replicas from Deployment php-apache-long-running
    not to have occurred‌

    /__w/1/s/gopath/src/github.com/Azure/aks-engine/test/e2e/kubernetes/kubernetes_test.go:1128
------------------------------‌
S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌S‌

Summarizing 1 Failure:‌

[Fail] ‌Azure Container Cluster using the Kubernetes Orchestrator ‌with a linux agent pool ‌[It] should be able to autoscale ‌
&#x2F;__w&#x2F;1&#x2F;s&#x2F;gopath&#x2F;src&#x2F;github.com&#x2F;Azure&#x2F;aks-engine&#x2F;test&#x2F;e2e&#x2F;kubernetes&#x2F;kubernetes_test.go:1128‌

Ran 32 of 47 Specs in 1995.981 seconds‌
FAIL!‌ -- ‌31 Passed‌ | ‌1 Failed‌ | ‌0 Pending‌ | ‌15 Skipped‌
--- FAIL: TestKubernetes (1995.98s)
FAIL
****
```

**Comment by PatrickLang** [comment]
Windows nodes are autoscaling successfully:

```
$ kubectl describe hpa iis-2019
Name:                                                  iis-2019
Namespace:                                             default
Labels:                                                <none>
Annotations:                                           <none>
CreationTimestamp:                                     Wed, 11 Sep 2019 22:41:30 +0000
Reference:                                             Deployment/iis-2019
Metrics:                                               ( current / target )
  resource cpu on pods  (as a percentage of request):  0% (0) / 10%
Min replicas:                                          2
Max replicas:                                          4
Deployment pods:                                       2 current / 2 desired
Conditions:
  Type            Status  Reason               Message
  ----            ------  ------               -------
  AbleToScale     True    ScaleDownStabilized  recent recommendations were higher than current one, applying the highest recent recommendation
  ScalingActive   True    ValidMetricFound     the HPA was able to successfully calculate a replica count from cpu resource utilization (percentage of request)
  ScalingLimited  False   DesiredWithinRange   the desired count is within the acceptable range
Events:
  Type     Reason                        Age                 From                       Message
  ----     ------                        ----                ----                       -------
  Normal   SuccessfulRescale             43m                 horizontal-pod-autoscaler  New size: 2; reason: Current number of replicas below Spec.MinReplicas
  Warning  FailedGetResourceMetric       40m (x12 over 43m)  horizontal-pod-autoscaler  unable to get metrics for resource cpu: no metrics returned from resource metrics API
  Warning  FailedComputeMetricsReplicas  40m (x12 over 43m)  horizontal-pod-autoscaler  invalid metrics (1 invalid out of 1), first error is: failed to get cpu utilization: unable to get metrics for resource cpu: no metrics returned from resource metrics API
  Normal   SuccessfulRescale             10m (x2 over 37m)   horizontal-pod-autoscaler  New size: 4; reason: cpu resource utilization (percentage of request) above target
  Normal   SuccessfulRescale             25s (x2 over 27m)   horizontal-pod-autoscaler  New size: 2; reason: All metrics below target

```


Linux failing in my manual tests:
```
$ kubectl describe hpa php-apache
Name:                                                  php-apache
Namespace:                                             default
Labels:                                                <none>
Annotations:                                           <none>
CreationTimestamp:                                     Wed, 11 Sep 2019 22:59:53 +0000
Reference:                                             Deployment/php-apache
Metrics:                                               ( current / target )
  resource cpu on pods  (as a percentage of request):  <unknown> / 30%
Min replicas:                                          2
Max replicas:                                          4
Deployment pods:                                       2 current / 2 desired
Conditions:
  Type           Status  Reason                   Message
  ----           ------  ------                   -------
  AbleToScale    True    SucceededGetScale        the HPA controller was able to get the target's current scale
  ScalingActive  False   FailedGetResourceMetric  the HPA was unable to compute the replica count: missing request for cpu
Events:
  Type     Reason                        Age                 From                       Message
  ----     ------                        ----                ----                       -------
  Normal   SuccessfulRescale             25m                 horizontal-pod-autoscaler  New size: 2; reason: Current number of replicas below Spec.MinReplicas
  Warning  FailedGetResourceMetric       23m (x11 over 25m)  horizontal-pod-autoscaler  unable to get metrics for resource cpu: no metrics returned from resource metrics API
  Warning  FailedComputeMetricsReplicas  23m (x11 over 25m)  horizontal-pod-autoscaler  invalid metrics (1 invalid out of 1), first error is: failed to get cpu utilization: unable to get metrics for resource cpu: no metrics returned from resource metrics API
  Warning  FailedComputeMetricsReplicas  22m                 horizontal-pod-autoscaler  invalid metrics (1 invalid out of 1), first error is: failed to get cpu utilization: missing request for cpu
  Warning  FailedGetResourceMetric       38s (x89 over 22m)  horizontal-pod-autoscaler  missing request for cpu

```

**Comment by PatrickLang** [comment]
Actually I had a mistake in my Linux deployment. I didn't have a request set, which needs to be there for metrics to be served per https://github.com/kubernetes/kubernetes/issues/79365#issuecomment-505682403

**Comment by PatrickLang** [comment]
At this point I need to stop work on this. Can someone else pick it up? I have higher priority stuff waiting on me and the other maintainers should be able to handle it from here. I see no blockers from a Windows standpoint and my manual tests are good.

File: pkg/api/defaults-kubelet.go [hunk 3]
```diff
@@ -117,6 +117,11 @@ func (cs *ContainerService) setKubeletConfig(isUpgrade bool) {
 		defaultKubeletConfig["--tls-cipher-suites"] = TLSStrongCipherSuitesKubelet
 	}
 
+	if common.IsKubernetesVersionGe(o.OrchestratorVersion, "1.16.0-beta.1") {
+		// for enabling metrics-server v0.3.0+
+		defaultKubeletConfig["--authentication-token-webhook"] = "true"
```
[the remark this answers is not in the document]
Comment (PatrickLang) [reply pkg/api/defaults-kubelet.go]: Manual tests on Windows were ok, resolving

**Comment by PatrickLang** [comment]
/hold cancel
Windows updates are in, no hold needed

**Pull request merged**